WalletWasabi.Tests/UnitTests/BlockchainAnalysis/CoinJoinAnonscoreTests.cs

```diff
@@ -0,0 +1,277 @@
+		}
+
+		[Fact]
+		public void MultiDenomination()
+		{
+			// Multiple standard denomination outputs results in correct calculation.
+			var analyser = ServiceFactory.CreateBlockchainAnalyzer();
+			var othersOutputs = new[] { 1, 1, 1, 2, 2 };
+			var tx = BitcoinFactory.CreateSmartTransaction(
+				9,
+				othersOutputs.Select(x => Money.Coins(x)),
+				new[] { (Money.Coins(3.2m), 1) },
+				new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet), (Money.Coins(2m), HdPubKey.DefaultHighAnonymitySet) });
+
+			analyser.Analyze(tx);
+
+			var level1 = tx.WalletOutputs.First(x => x.Amount == Money.Coins(1m));
+			var level2 = tx.WalletOutputs.First(x => x.Amount == Money.Coins(2m));
+
+			Assert.Equal(1, tx.WalletInputs.First().HdPubKey.AnonymitySet);
+			Assert.Equal(4, level1.HdPubKey.AnonymitySet);
+			Assert.Equal(3, level2.HdPubKey.AnonymitySet);
+		}
+
+		[Fact]
+		public void MultiDenominationInheritance()
+		{
+			// Multiple denominations inherit properly.
+			var analyser = ServiceFactory.CreateBlockchainAnalyzer();
+			var othersOutputs = new[] { 1, 1, 1, 2, 2 };
+			var tx = BitcoinFactory.CreateSmartTransaction(
+				9,
+				othersOutputs.Select(x => Money.Coins(x)),
+				new[] { (Money.Coins(3.2m), 100) },
+				new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet), (Money.Coins(2m), HdPubKey.DefaultHighAnonymitySet) });
+
+			analyser.Analyze(tx);
+
+			var level1 = tx.WalletOutputs.First(x => x.Amount == Money.Coins(1m));
+			var level2 = tx.WalletOutputs.First(x => x.Amount == Money.Coins(2m));
+
+			Assert.Equal(100, tx.WalletInputs.First().HdPubKey.AnonymitySet);
+			Assert.Equal(103, level1.HdPubKey.AnonymitySet);
+			Assert.Equal(102, level2.HdPubKey.AnonymitySet);
+		}
+
+		[Fact]
+		public void SelfAnonsetSanityCheck()
+		{
+			// If we have multiple same denomination in the same coinjoin, then don't gain anonset on ourselves.
+			var analyser = ServiceFactory.CreateBlockchainAnalyzer();
+			var othersOutputs = new[] { 1, 1, 1 };
+			var ownOutputs = new[] { 1, 1 };
+			var tx = BitcoinFactory.CreateSmartTransaction(
+				9,
+				othersOutputs.Select(x => Money.Coins(x)),
+				new[] { (Money.Coins(3.2m), 1) },
+				ownOutputs.Select(x => (Money.Coins(x), HdPubKey.DefaultHighAnonymitySet)));
+
+			analyser.Analyze(tx);
+
+			Assert.Equal(1, tx.WalletInputs.First().HdPubKey.AnonymitySet);
+			Assert.All(tx.WalletOutputs, x => Assert.Equal(4, x.HdPubKey.AnonymitySet));
+		}
+
+		[Fact]
+		public void InputSanityCheck()
+		{
+			// Anonset can never be larger than the number of inputs.
+			var analyser = ServiceFactory.CreateBlockchainAnalyzer();
+			var tx = BitcoinFactory.CreateSmartTransaction(2, Enumerable.Repeat(Money.Coins(1m), 9), new[] { (Money.Coins(1.1m), 1) }, new[] { (Money.Coins(1m), HdPubKey.DefaultHighAnonymitySet) });
+
+			analyser.Analyze(tx);
+
+			Assert.Equal(1, tx.WalletInputs.First().HdPubKey.AnonymitySet);
+			Assert.Equal(3, tx.WalletOutputs.First().HdPubKey.AnonymitySet);
+		}
```

Comment:
i'm confused by this test, not sure what it's asserting really

Reply:
"Anonset can never be larger than the number of inputs."

Assume you see a coinjoin where 10 outputs are equal, but there are only 3 inputs. In this case the outputs' anonset shouldn't be 10, but at maximum 3 as at maximum 3 people could have possibly participate in the transaction.